crates/bevy_winit/src/lib.rs

```diff
@@ -922,9 +880,24 @@ pub fn winit_runner(mut app: App) {
     };
 
     trace!("starting winit event loop");
-    if return_from_run {
-        run_return(&mut event_loop, event_handler);
-    } else {
-        run(event_loop, event_handler);
+    if let Err(err) = event_loop.run(event_handler) {
+        error!("winit event loop returned an error: {err}");
     }
 }
+
+fn react_to_resize(
+    window: &mut Mut<'_, Window>,
+    size: winit::dpi::PhysicalSize<u32>,
+    event_writers: &mut WindowAndInputEventWriters<'_>,
+    window_entity: Entity,
+) {
+    window
+        .resolution
+        .set_physical_resolution(size.width, size.height);
+
+    event_writers.window_resized.send(WindowResized {
+        window: window_entity,
+        width: window.width(),
+        height: window.height(),
+    });
```

Comment:
Commenting these lines "fix" https://github.com/bevyengine/bevy/pull/10702#issuecomment-1826829623. (well it breaks the resizing but it doesn't crash.) @extrawurst 

The problem seems to be the computation of height/width, doubling multiple times until reaching the threshold for main_texture_a.

hypothesis: when a resize occurs (it seems to occur several times on web (due to adapting to window, and canvas size I guess, gotta check those out.), the size is computed there, and maybe its missing a scale factor, or not using correct properties 🤔  ; not sure.

Reply:
I can reproduce the problem on mac (firefox) ; doesn't seem to occur in windows (firefox, chrome)

Reply:
https://github.com/bevyengine/bevy/issues/7038 could maybe be related?

Reply:
What's surprising to me is bevy doesn't call `request_inner_size` 🤔.

<details><summary>Sanity check</summary>
<p>

- I initially added logs, and not seeing [those](https://github.com/Vrixyz/bevy/blob/2cc695441fad367337c9ac1a92b27716a457799b/crates/bevy_winit/src/system.rs#L179) [logs](https://github.com/Vrixyz/bevy/blob/2cc695441fad367337c9ac1a92b27716a457799b/crates/bevy_winit/src/lib.rs#L581) was suspicious.
- I tested my assumption by adding a panic!(), in case my log was somehow swallowed: It didn't trigger.
</p>
</details> 

Concerning `CanvasParentResizePlugin`, it might be interesting to discuss:
on the logs I shared, it doesn't impact anything, because [it's not enabled](https://github.com/Vrixyz/bevy/blob/2cc695441fad367337c9ac1a92b27716a457799b/crates/bevy_winit/src/system.rs#L95), as `fit_canvas_to_parent` is not enabled.

Now, if I enable it, the crash is gone. (thanks @extrawurst for pointing me out to [bevy_wasm_window_resize](https://github.com/Leinnan/bevy_wasm_window_resize/blob/master/src/lib.rs) which fixes it too).

<details><summary>Logs with `fit_canvas_to_parent = true`</summary>
<p>

```rs
 INFO app: crates/bevy_winit/src/system.rs:55 Creating new window "I am a window!" (0v0)
 INFO app: crates/bevy_winit/src/lib.rs:367 NewEvents(Init)
 INFO app: crates/bevy_winit/src/lib.rs:367 Resumed
 INFO app: crates/bevy_winit/src/lib.rs:367 AboutToWait
 INFO app: crates/bevy_render/src/renderer/mod.rs:140 AdapterInfo { name: "Apple M1", vendor: 4203, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Gl }
 INFO app: crates/bevy_winit/src/lib.rs:367 NewEvents(WaitCancelled { start: Instant(2.59292s), requested_resume: None })
 WARN app: crates/bevy_pbr/src/ssao/mod.rs:87 ScreenSpaceAmbientOcclusionPlugin not loaded. GPU lacks support: TextureFormat::R16Float does not support TextureUsages::STORAGE_BINDING.
 INFO app: crates/bevy_winit/src/lib.rs:367 WindowEvent { window_id: WindowId(WindowId(1)), event: RedrawRequested }
 INFO app: crates/bevy_winit/src/lib.rs:367 AboutToWait
  // (note added for github discussion): here we see our log:
 INFO app: crates/bevy_winit/src/web_resize.rs:35 canvas_parent_resize_event_handler: LogicalSize { width: 1428.0, height: 774.0 }
 INFO app: crates/bevy_render/src/view/mod.rs:440 prepare_view_targets
 INFO app: crates/bevy_render/src/view/mod.rs:448 UVec2(2560, 1440)
 INFO app: crates/bevy_winit/src/lib.rs:367 NewEvents(Poll)
 INFO app: crates/bevy_winit/src/lib.rs:367 WindowEvent { window_id: WindowId(WindowId(1)), event: Resized(PhysicalSize { width: 2856, height: 1548 }) }
 INFO app: crates/bevy_winit/src/lib.rs:471 WindowEvent::Resized sent us: PhysicalSize { width: 2856, height: 1548 }
 INFO app: crates/bevy_winit/src/lib.rs:895 react_to_resize: PhysicalSize { width: 2856, height: 1548 }
 INFO app: crates/bevy_winit/src/lib.rs:367 WindowEvent { window_id: WindowId(WindowId(1)), event: RedrawRequested }
 INFO app: crates/bevy_winit/src/lib.rs:367 AboutToWait
 INFO app: crates/bevy_render/src/view/mod.rs:440 prepare_view_targets
 INFO app: crates/bevy_render/src/view/mod.rs:448 UVec2(2856, 1548)
 INFO app: crates/bevy_winit/src/lib.rs:367 NewEvents(Poll)
 INFO app: crates/bevy_winit/src/lib.rs:367 AboutToWait
 INFO app: crates/bevy_render/src/view/mod.rs:440 prepare_view_targets
 INFO app: crates/bevy_render/src/view/mod.rs:448 UVec2(2856, 1548)
 ```

</p>
</details> 

I didn't succeed in reproducing an issue with winit examples.

My next lead is bevy is missing some initial setup for winit, where we should inform winit of the initial canvas size 🤔.

Reply:
I think the bug comes from a misunderstanding on my end on differences between logical size, physical size, scaling factor, and mapping between `bevy_winit` and `bevy_windows`. I'm 90% sure that's not a `winit` bug :tada:

The crash can be fixed with [this snippet](https://github.com/Vrixyz/bevy/pull/5/commits/467fb07943449d84a7353f907a96fb32d57b2670#diff-b1447f30819280931099e7498b2ced388a98de5cda1b631b493107bb96193cc0R898-R908): dividing (physical) `size` per `scale_factor`. But I'd like to understand it fully.

<details><summary>Logs</summary>
<p>

  INFO app: crates/bevy_winit/src/system.rs:55 Creating new window "I am a window!" (0v0)
 INFO app: crates/bevy_render/src/renderer/mod.rs:140 AdapterInfo { name: "Apple M1", vendor: 4203, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Gl }
 WARN app: crates/bevy_pbr/src/ssao/mod.rs:87 ScreenSpaceAmbientOcclusionPlugin not loaded. GPU lacks support: TextureFormat::R16Float does not support TextureUsages::STORAGE_BINDING.
 INFO app: crates/bevy_render/src/view/mod.rs:440 prepare_view_targets
 INFO app: crates/bevy_render/src/view/mod.rs:448 UVec2(2560, 1440)
 INFO app:  size_handler: PhysicalSize { width: 5120, height: 2880 } log.target = "winit::platform_impl::platform::event_loop::window_target";
log.module_path = "winit::platform_impl::platform::event_loop::window_target";
log.file = "/Users/thierryberger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.4/src/platform_impl/web/event_loop/window_target.rs";
log.line = 619;
 INFO app: crates/bevy_winit/src/lib.rs:471 WindowEvent::Resized sent us: PhysicalSize { width: 5120, height: 2880 }
 INFO app: crates/bevy_winit/src/lib.rs:896 react_to_resize: PhysicalSize { width: 5120, height: 2880 }
 INFO app: crates/bevy_winit/src/lib.rs:897 window: (1280.0, 720.0)
 INFO app: crates/bevy_winit/src/lib.rs:899 scale_factor: 2
 INFO app: crates/bevy_winit/src/lib.rs:904 new_resolution: (2560.0, 1440.0)
 INFO app: crates/bevy_winit/src/lib.rs:909 new window: (1280.0, 720.0)
 INFO app: crates/bevy_render/src/view/mod.rs:440 prepare_view_targets
 INFO app: crates/bevy_render/src/view/mod.rs:448 UVec2(2560, 1440)
 INFO app: crates/bevy_render/src/view/mod.rs:440 prepare_view_targets
 INFO app: crates/bevy_render/src/view/mod.rs:448 UVec2(2560, 1440)

</p>
</details> 

<details><summary>canvas screenshot</summary>
<p>

<img width="1330" alt="bevy winit 0 29 wasm canvas size" src="https://github.com/bevyengine/bevy/assets/2290685/9cc6887c-7e24-4c94-bd13-2e7b3f1fac4a">

</p>
</details> 

What's weird to me, is the difference between different logged sizes:
- the default value for our physical width [is **1280**](https://github.com/Vrixyz/bevy/blob/467fb07943449d84a7353f907a96fb32d57b2670/crates/bevy_window/src/window.rs#L624)
- width reported by the browser to ResizeObserver to winit to bevy: **5120**: that's our physical width * 2 * 2 🤔 

My lead is that there is a scale applied (or not applied?) at the wrong place somewhere 🕵️🤔. 

[That line seems](https://github.com/Vrixyz/bevy/blob/467fb07943449d84a7353f907a96fb32d57b2670/crates/bevy_window/src/window.rs#L636) suspicious to me, mapping logical directly to physical ? That's a bit confusing.

Reply:
I think indeed that’s a recent change from winit, where it used to send us the physical size without scale applied, but now it is applied, so [that line](https://github.com/bevyengine/bevy/pull/10702/files#diff-b1447f30819280931099e7498b2ced388a98de5cda1b631b493107bb96193cc0L523) (I copied in react_to_resize) should cancel the scale, or set the logical size rather than physical (similar end result).

That still doesn't explain why the browser also applies the scale twice, there must be somewhere else impacted.

Reply:
That behaviour you described seems correct.

Also, when I resize the canvas via web debugger, the behaviour seems fine.

What I'm left to understand is where my canvas width is set to 2560, which I guess provokes a resize event from browser to physical width 5120. Where is it asked for..? That resize seems to happen after 1 frames...

Reply:
Considering that the new version of Winit doesn't touch the canvas size at all unless requested through `Window::request_inner_size()` or `InnerSizeWriter::request_inner_size()`, it should be Bevy doing it.

The problem might be Wgpu, where `Surface::configure()` will change the `width` and `height` of the canvas, which takes logical size. But this was done back in v0.16, so I'm not sure why you are hitting that problem just now.

Generally I would recommend Bevy to set the CSS size of the canvas to something, so that `Surface::configure()` has no visual effect.

Reply:
This is a bit confusing so bear with me here.

There are two different sizes you have to take into account:
- Output Bitmap Size: the size of the surface you are going to draw on.
- Canvas Size: the size of the canvas as it is seen on the screen.

The canvas `width` and `height` attribute sets the output bitmap size in physical pixels, but also the canvas size in logical pixels. You can see how this creates an undesirable effect.

The solution: to set the canvas size in CSS. This will make the canvas **not** use the `width` and `height` attribute for its size, but continue to use it for the output bitmap size.
Winit will report to you the true size of the canvas in physical pixels that you then use to set the canvas `width` and `height` attribute through `Surface::configure()`, but now that you have set a CSS size on the canvas, it should stop having this weird back and forth.

Reply:
Setting the size in the canvas indeed fixes it! Thanks for following me on that in depth analysis!